MRG: loading TCL / Tk symbols dynamically

matthew-brett opened the pull request:
This is an attempt to load the symbols we need from the Tkinter.tkinter
module at run time, rather than by linking at build time.

It is one way of building a manylinux wheel that can build on a basic
Manylinux docker image, and then run with the TCL / Tk library installed
on the installing user's machine.  It would also make the situation
better on OSX, where we have to build against ActiveState TCL for
compatibility with Python.org Python, but we would like to allow
run-time TCL from e.g. homebrew.

I have tested this on Debian Jessie Python 2.7 and 3.5, and on OSX 10.9
with Python 2.7.

Questions:

* Would y'all consider carrying something like this approach in
  the matplotlib source, but not enabled by default, to help building
  binary wheels?
* Do you have any better suggestions about how to do this?
* My C fu is weak; is there a way of collecting the typedefs I need from
  the TCL / Tk headers rather than copying them into the _tkagg.cpp
  source file (typdefs starting around line 52)?
* My fu for Python C extension modules is also weak; did I configure
  exceptions and handle references correctly?

mdboom commented:
Thanks for this.  This is really helpful.

@matthew-brett wrote:
> Would y'all consider carrying something like this approach in the matplotlib source, but not enabled by default, to help building binary wheels?

I'd prefer to give this loads of testing and just have this as the "one way to do it" if we can.  We could eliminate tons of hacks in the build script and tons of pain and support using this approach, at the expense of the very minor dynamic linking at import time.

> My C fu is weak; is there a way of collecting the typedefs I need from the TCL / Tk headers rather than copying them into the _tkagg.cpp source file (typdefs starting around line 52)?

I don't think so.

> My fu for Python C extension modules is also weak; did I configure exceptions and handle references correctly?

I'll read through and try to verify.

mdboom commented on src/_tkagg.cpp:
```diff
@@ -232,6 +264,90 @@ static PyMethodDef functions[] = {
+#define TKINTER_PKG "tkinter"
+#define TKINTER_MOD "_tkinter"
+// From module __file__ attribute to char *string for dlopen.
+#define FNAME2CHAR(s) (PyBytes_AsString(PyUnicode_EncodeFSDefault(s)))
```
This isn't exception proof -- `PyUnicode_EncodeFSDefault` could raise an exception here.  Maybe just make a regular old function for this purpose?

mdboom commented on src/_tkagg.cpp:
```diff
@@ -232,6 +264,90 @@ static PyMethodDef functions[] = {
+        // reset errors
+        dlerror();
+        func = dlsym(lib_handle, func_name);
+        const char *error = dlerror();
```
I think this only needs to be called if `func == NULL`.

mdboom commented on src/_tkagg.cpp:
```diff
@@ -232,6 +264,90 @@ static PyMethodDef functions[] = {
+    // Load function, unless there has been a previous error.  If so, then
+    // return NULL.  If there is an error loading the function, return NULL
+    // and set error flag.
+    static int have_error = 0;
```
This is fine, but is a sort of unidiomatic pattern.  Rather than using a static variable `has_error` here, why not just return NULL if there was an error and then call it differently from `_func_loader`, e.g.:

```
if (!((TCL_CREATE_COMMAND = (tcl_cc) _dfunc(tkinter_lib, "Tcl_CreateCommand")) &&
      (TCL_APPEND_RESULT = (tcl_cc) _dfunc(tkinter_lib, "Tcl_AppendResult")) &&
    ... etc ..
```

mdboom commented on src/_tkagg.cpp:
```diff
@@ -232,6 +264,90 @@ static PyMethodDef functions[] = {
+    int ret = -1;
+    PyObject *pModule, *pSubmodule, *pString;
+
+    pModule = PyImport_ImportModule(TKINTER_PKG);
```
It's a matter of personal preference, but I generally like to avoid the excessive nesting by:

1) setting all `PyObject *` to NULL in their declaration
2) doing:

```
if (pModule == NULL) {
  goto exit;
}
```

3) having

```
exit:
   Py_XDECREF(pString);
   Py_XDECREF(pSubmodule);
   Py_XDECREF(pModule);

   return ret;
```

at the bottom

QuLogic commented:
> My C fu is weak; is there a way of collecting the typedefs I need from the TCL / Tk headers rather than copying them into the _tkagg.cpp source file (typdefs starting around line 52)?

What's wrong with `#include`ing the header?

mdboom commented:
> What's wrong with #includeing the header?

Because the point of this change is to replace actual direct function calls with function pointers.  Numpy, for example, has a special generated header for this exact purpose (`__multiarray_api.h`), but most C libraries don't.

QuLogic commented:
I was hoping since it's C++ you could use some `auto` magic, but that might be C++11 only.

dopplershift commented:
`auto` is static (i.e. compile-time) typing only, so there's no way that could work for runtime type inference. That's the realm of C++ virtual functions/polymorphism.

QuLogic commented:
We don't need runtime typing; we just need the type of the function at compile time so that we can set up a pointer to it.

dopplershift commented:
The return type of `dlsym()` (and hence `_dfunc`) is `void *`; there is absolutely no way for the compiler to help you out here.

QuLogic commented:
I didn't really mean `auto`, just some "C++ magic" like it. And that something would be `typeof` or `decltype`, but both of these come with their own issues (one's a GNU-ism, the other's C++11 only).

ogrisel commented on src/_tkagg.cpp:
```diff
@@ -232,6 +264,90 @@ static PyMethodDef functions[] = {
+#ifdef DYNAMIC_TKINTER
+// Functions to fill global TCL / Tk function pointers from tkinter module.
+
+#include <dlfcn.h>
```
This include fails when using VS

```
src/_tkagg.cpp(270) : fatal error C1083: Cannot open include file: 'dlfcn.h': No such file or directory
```

https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1545/job/724qv5hcfngqa2r9#L1033

ogrisel replied:
Actually more recent versions of Visual C++ also fail, e.g. in the Python 3.5 Windows build on AppVeyor:

```
src/_tkagg.cpp(270): fatal error C1083: Cannot open include file: 'dlfcn.h': No such file or directory 
error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\amd64\\cl.exe' failed with exit status 2
Command exited with code 1
```

https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1545/job/rk6safsaid63i7ch#L1053

matthew-brett commented:
@mdboom - addressed your comments - I think.  I've tested disabling the linker flags for TCL / Tk libraries in `setupext.py` - works on OSX Python 2.7, 3.5, Debian Jessie 2.7 and 3.5.

mdboom commented:
Cool.  If this works, we might as well take most of the crazy header-file-hunting out of `setupext.py` as well.

As for the Windows-specific issues, @cgohlke, any thoughts?

matthew-brett commented:
I've done some speculative changes to support Windows, trying to work out how to test them now.

matthew-brett commented:
OK - Windows changes are compiling and passing tests on Python 2.7 at least.

I propose to make remove the not-dynamic branch, and disable linking the TCL / Tk libs in `setupext.py` to see if this can work without build-time linking.

matthew-brett commented:
Build-time linking disabled, Appveyor tests passing, failure on travis I believe is unrelated, so from my point of view, I think this is ready.

cgohlke commented:
Maybe I am missing something, but how can this work on Windows where the Tcl/Tk symbols are not exported from `_tkinter.pyd`? Is appveyor testing TkAgg?

```Python
>>> from ctypes import *
>>> from ctypes.wintypes import *
>>> windll.kernel32.LoadLibraryW.restype = HMODULE
>>> windll.kernel32.LoadLibraryW.argtypes = [LPCWSTR]
>>> windll.kernel32.GetProcAddress.argtypes = [HMODULE, LPCSTR]
>>> windll.kernel32.GetProcAddress.restype = c_void_p
>>> from tkinter import _tkinter
>>> hdll = windll.kernel32.LoadLibraryW(_tkinter.__file__)
>>> print(windll.kernel32.GetProcAddress(hdll, b"PyInit__tkinter"))
140732219092336
>>> print(windll.kernel32.GetProcAddress(hdll, b"Tcl_CreateCommand"))
None
```

efiring commented:
Quick test with a conda python 3.5 environment on OS X 10.9.5: it works!  I was not able to get a working Tk backend without this.
Maybe unrelated to this, but I get lots of warnings while executing:
```
2016-05-18 10:51:04.230 python[18741:507] setCanCycle: is deprecated.  Please use setCollectionBehavior instead
```

matthew-brett commented:
Christoph - sorry - I'm behind a fierce proxy in Cuba, and I can't compile myself.  Google is blocking me from downloading the Appveyor build artifacts with `We're sorry, but this service is not available in your location` (Google code blocks Cuba too).

The code at https://github.com/matplotlib/matplotlib/pull/6442/files#diff-c4f5727f595ec7b41e86800bb717a0efR323 should be checking for null pointers at run time, and raising an exception.

What do you get if you try and make and show a plot with the tkagg backend using one the build artifacts on Appveyor - e.g. https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1553/job/c4lm62wyn1541509/artifacts ?

cgohlke commented:
Using `matplotlib-1.5.1+1973.g7257404-cp27-cp27m-win_amd64.whl`:
```Python
>>> import matplotlib
>>> matplotlib.use('TkAgg')
>>> matplotlib.__version__
'1.5.1+1973.g7257404'
>>> from matplotlib import pyplot
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "matplotlib\pyplot.py", line 113, in <module>
    _backend_mod, new_figure_manager, draw_if_interactive, _show = pylab_setup()
  File "matplotlib\backends\__init__.py", line 55, in pylab_setup
    [backend_name], 0)
  File "matplotlib\backends\backend_tkagg.py", line 13, in <module>
    import matplotlib.backends.tkagg as tkagg
  File "matplotlib\backends\tkagg.py", line 9, in <module>
    from matplotlib.backends import _tkagg
RuntimeError: Cannot load function Tcl_CreateCommand
```
Also, the artifact is missing zlib.dll:
```Python
>>> from matplotlib import pyplot
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "matplotlib\pyplot.py", line 29, in <module>
    import matplotlib.colorbar
  File "matplotlib\colorbar.py", line 34, in <module>
    import matplotlib.collections as collections
  File "matplotlib\collections.py", line 27, in <module>
    import matplotlib.backend_bases as backend_bases
  File "matplotlib\backend_bases.py", line 63, in <module>
    import matplotlib.textpath as textpath
  File "matplotlib\textpath.py", line 20, in <module>
    from matplotlib.mathtext import MathTextParser
  File "matplotlib\mathtext.py", line 62, in <module>
    import matplotlib._png as _png
ImportError: DLL load failed: The specified module could not be found.
```

matthew-brett commented:
The png failure must be unrelated.   I suppose the GetProcAddress failure is related to [this](https://en.wikipedia.org/wiki/Dynamic_loading#Special_Library)

    Windows allows programmers to access symbols exported by the main executable. 
    Windows does not use a global symbol table, and has no API to search across multiple
    modules to find a symbol by name.

So, the next question is - do you (Christoph) or does anyone know how to query `_tkinter.pyd` at run time for its library dependencies?

cgohlke commented:
Using [pefile](https://github.com/erocarrera/pefile/):
```Python
>>> import pefile
>>> from tkinter import _tkinter
>>> pe = pefile.PE(_tkinter.__file__)
>>> for entry in pe.DIRECTORY_ENTRY_IMPORT:
...     print(entry.dll)
...     for imp in entry.imports:
...         print('\t', hex(imp.address), imp.name)
...
b'tcl86t.dll'
         0x180009458 b'Tcl_ThreadQueueEvent'
         0x180009460 b'TclBN_mp_clear'
         0x180009468 b'Tcl_ListObjIndex'
         0x180009470 b'TclBN_mp_init'
         0x180009478 b'Tcl_GetVar2'
         0x180009480 b'Tcl_Init'
         0x180009488 b'TclBN_mp_unsigned_bin_size'
         0x180009490 b'Tcl_SetVar2'
         0x180009498 b'Tcl_ExprString'
         0x1800094a0 b'Tcl_UnsetVar2'
         0x1800094a8 b'Tcl_Eval'
         0x1800094b0 b'Tcl_DeleteCommand'
         0x1800094b8 b'Tcl_NewWideIntObj'
         0x1800094c0 b'Tcl_NewByteArrayObj'
         0x1800094c8 b'Tcl_NewLongObj'
         0x1800094d0 b'Tcl_RecordAndEval'
         0x1800094d8 b'Tcl_GetUnicode'
         0x1800094e0 b'Tcl_GetVar2Ex'
         0x1800094e8 b'Tcl_GetString'
         0x1800094f0 b'Tcl_NewUnicodeObj'
         0x1800094f8 b'Tcl_GetDouble'
         0x180009500 b'Tcl_GetBoolean'
         0x180009508 b'Tcl_AttemptAlloc'
         0x180009510 b'TclBN_mp_read_radix'
         0x180009518 b'Tcl_ListObjLength'
         0x180009520 b'Tcl_ThreadAlert'
         0x180009528 b'Tcl_GetBooleanFromObj'
         0x180009530 b'Tcl_AddErrorInfo'
         0x180009538 b'Tcl_EvalObjv'
         0x180009540 b'Tcl_FindExecutable'
         0x180009548 b'Tcl_ConditionWait'
         0x180009550 b'Tcl_ListObjGetElements'
         0x180009558 b'Tcl_SetObjResult'
         0x180009560 b'Tcl_NewDoubleObj'
         0x180009568 b'Tcl_DeleteInterp'
         0x180009570 b'Tcl_GetCurrentThread'
         0x180009578 b'Tcl_MutexUnlock'
         0x180009580 b'Tcl_ConditionNotify'
         0x180009588 b'Tcl_GetObjType'
         0x180009590 b'Tcl_GetObjResult'
         0x180009598 b'Tcl_ExprBoolean'
         0x1800095a0 b'Tcl_GetBignumFromObj'
         0x1800095a8 b'Tcl_GetThreadData'
         0x1800095b0 b'Tcl_GetStringResult'
         0x1800095b8 b'Tcl_NewIntObj'
         0x1800095c0 b'Tcl_GetCharLength'
         0x1800095c8 b'Tcl_NewStringObj'
         0x1800095d0 b'Tcl_CreateCommand'
         0x1800095d8 b'Tcl_MutexLock'
         0x1800095e0 b'Tcl_NewListObj'
         0x1800095e8 b'TclBN_mp_to_unsigned_bin_n'
         0x1800095f0 b'Tcl_ResetResult'
         0x1800095f8 b'Tcl_DeleteTimerHandler'
         0x180009600 b'Tcl_ConditionFinalize'
         0x180009608 b'Tcl_CreateInterp'
         0x180009610 b'Tcl_GetWideIntFromObj'
         0x180009618 b'Tcl_NewBignumObj'
         0x180009620 b'Tcl_CreateTimerHandler'
         0x180009628 b'Tcl_ExprDouble'
         0x180009630 b'Tcl_DoOneEvent'
         0x180009638 b'Tcl_SetVar2Ex'
         0x180009640 b'Tcl_GetDoubleFromObj'
         0x180009648 b'Tcl_SplitList'
         0x180009650 b'Tcl_EvalFile'
         0x180009658 b'Tcl_Free'
         0x180009660 b'Tcl_GetByteArrayFromObj'
         0x180009668 b'Tcl_GetLongFromObj'
         0x180009670 b'TclFreeObj'
         0x180009678 b'Tcl_ExprLong'
         0x180009680 b'Tcl_GetStringFromObj'
b'tk86t.dll'
         0x180009690 b'Tk_GetNumMainWindows'
         0x180009698 b'Tk_MainWindow'
         0x1800096a0 b'Tk_Init'
b'KERNEL32.dll'
         0x180009000 b'GetLastError'
         0x180009008 b'GetModuleHandleW'
         0x180009010 b'GetStartupInfoW'
         0x180009018 b'IsDebuggerPresent'
         0x180009020 b'InitializeSListHead'
         0x180009028 b'DisableThreadLibraryCalls'
         0x180009030 b'GetSystemTimeAsFileTime'
         0x180009038 b'GetCurrentThreadId'
         0x180009040 b'GetCurrentProcessId'
         0x180009048 b'QueryPerformanceCounter'
         0x180009050 b'IsProcessorFeaturePresent'
         0x180009058 b'TerminateProcess'
         0x180009060 b'GetCurrentProcess'
         0x180009068 b'SetUnhandledExceptionFilter'
         0x180009070 b'UnhandledExceptionFilter'
         0x180009078 b'RtlVirtualUnwind'
         0x180009080 b'RtlLookupFunctionEntry'
         0x180009088 b'RtlCaptureContext'
         0x180009090 b'Sleep'
         0x180009098 b'GetEnvironmentVariableW'
         0x1800090a0 b'SetEnvironmentVariableW'
b'python35.dll'
         0x180009170 b'Py_GetProgramName'
         0x180009178 b'PyUnicode_FromString'
         0x180009180 b'PyErr_Print'
         0x180009188 b'_Py_TrueStruct'
         0x180009190 b'_PyLong_Format'
         0x180009198 b'PyMem_Malloc'
         0x1800091a0 b'PyCallable_Check'
         0x1800091a8 b'_PyUnicode_Ready'
         0x1800091b0 b'PyNumber_Float'
         0x1800091b8 b'PyObject_Str'
         0x1800091c0 b'PyObject_IsTrue'
         0x1800091c8 b'PyExc_TypeError'
         0x1800091d0 b'_PyObject_New'
         0x1800091d8 b'PyNumber_Check'
         0x1800091e0 b'PyType_FromSpec'
         0x1800091e8 b'_Py_NotImplementedStruct'
         0x1800091f0 b'PyBytes_FromStringAndSize'
         0x1800091f8 b'PyUnicode_FromKindAndData'
         0x180009200 b'PyErr_CheckSignals'
         0x180009208 b'PyErr_NoMemory'
         0x180009210 b'PyNumber_Negative'
         0x180009218 b'PyTuple_SetItem'
         0x180009220 b'PyLong_AsLongAndOverflow'
         0x180009228 b'PyLong_FromVoidPtr'
         0x180009230 b'PyMem_Free'
         0x180009238 b'PyThread_free_lock'
         0x180009240 b'PyBytes_Size'
         0x180009248 b'PyTuple_Size'
         0x180009250 b'PyErr_Clear'
         0x180009258 b'PySequence_Tuple'
         0x180009260 b'PyEval_RestoreThread'
         0x180009268 b'PyEval_CallObjectWithKeywords'
         0x180009270 b'PyUnicode_AsUTF8String'
         0x180009278 b'PySequence_Size'
         0x180009280 b'_PyArg_Parse_SizeT'
         0x180009288 b'PyThreadState_Get'
         0x180009290 b'PyErr_SetObject'
         0x180009298 b'PyBool_FromLong'
         0x1800092a0 b'_PyObject_CallFunction_SizeT'
         0x1800092a8 b'PyUnicode_EncodeFSDefault'
         0x1800092b0 b'PyModule_AddIntConstant'
         0x1800092b8 b'_PyErr_BadInternalCall'
         0x1800092c0 b'PyBytes_AsString'
         0x1800092c8 b'_PyLong_AsByteArray'
         0x1800092d0 b'PyErr_NormalizeException'
         0x1800092d8 b'PyErr_Occurred'
         0x1800092e0 b'PyUnicode_DecodeUTF8'
         0x1800092e8 b'PyObject_GenericGetAttr'
         0x1800092f0 b'PyUnicode_AsUTF8AndSize'
         0x1800092f8 b'PyEval_SaveThread'
         0x180009300 b'PyExc_RuntimeError'
         0x180009308 b'_Py_ctype_table'
         0x180009310 b'PyLong_FromLong'
         0x180009318 b'PyUnicode_Concat'
         0x180009320 b'PyThread_allocate_lock'
         0x180009328 b'Py_GetPrefix'
         0x180009330 b'_Py_NoneStruct'
         0x180009338 b'PyTuple_New'
         0x180009340 b'PyLong_FromLongLong'
         0x180009348 b'PyThread_acquire_lock'
         0x180009350 b'PyFloat_FromDouble'
         0x180009358 b'PyUnicode_FromWideChar'
         0x180009360 b'PyErr_SetString'
         0x180009368 b'_PyTuple_Resize'
         0x180009370 b'PyExc_ValueError'
         0x180009378 b'PyErr_Format'
         0x180009380 b'PyModule_AddStringConstant'
         0x180009388 b'PyBool_Type'
         0x180009390 b'_Py_FalseStruct'
         0x180009398 b'PyFloat_Type'
         0x1800093a0 b'_PyLong_FromByteArray'
         0x1800093a8 b'PyLong_Type'
         0x1800093b0 b'PyType_IsSubtype'
         0x1800093b8 b'PyErr_Restore'
         0x1800093c0 b'PyExc_OverflowError'
         0x1800093c8 b'PyTuple_GetItem'
         0x1800093d0 b'PyObject_Free'
         0x1800093d8 b'PyErr_ExceptionMatches'
         0x1800093e0 b'PyThread_release_lock'
         0x1800093e8 b'PyExc_UnicodeDecodeError'
         0x1800093f0 b'_Py_ctype_tolower'
         0x1800093f8 b'PyModule_AddObject'
         0x180009400 b'_PyArg_ParseTuple_SizeT'
         0x180009408 b'PyErr_Fetch'
         0x180009410 b'PyUnicode_AsUTF8'
         0x180009418 b'PyUnicode_AsWideCharString'
         0x180009420 b'PyUnicode_FromFormat'
         0x180009428 b'PyErr_BadArgument'
         0x180009430 b'_Py_stat'
         0x180009438 b'PyModule_Create2'
         0x180009440 b'PyOS_InputHook'
         0x180009448 b'PyErr_NewException'
b'VCRUNTIME140.dll'
         0x1800090b0 b'memchr'
         0x1800090b8 b'__telemetry_main_invoke_trigger'
         0x1800090c0 b'__telemetry_main_return_trigger'
         0x1800090c8 b'__C_specific_handler'
         0x1800090d0 b'__std_type_info_destroy_list'
         0x1800090d8 b'memset'
b'api-ms-win-crt-stdio-l1-1-0.dll'
         0x180009150 b'_kbhit'
b'api-ms-win-crt-runtime-l1-1-0.dll'
         0x1800090e8 b'_seh_filter_dll'
         0x1800090f0 b'_initterm_e'
         0x1800090f8 b'_initterm'
         0x180009100 b'_errno'
         0x180009108 b'_initialize_narrow_environment'
         0x180009110 b'_initialize_onexit_table'
         0x180009118 b'_register_onexit_function'
         0x180009120 b'_execute_onexit_table'
         0x180009128 b'_crt_atexit'
         0x180009130 b'_crt_at_quick_exit'
         0x180009138 b'_cexit'
         0x180009140 b'terminate'
b'api-ms-win-crt-string-l1-1-0.dll'
         0x180009160 b'strcmp'
```

matthew-brett commented:
Rather than analyze the DLL file and add an extra dependency, I am trying to search through the loaded modules in the current process, looking for the TCL / Tk routines.  This seems to be very fast on simple tests.

matthew-brett commented:
Christoph - would you mind checking https://ci.appveyor.com/project/mdboom/matplotlib/build/1.0.1558/job/snxq695d8h4sgqli/artifacts

matthew-brett commented:
I finally got a Windows test install working - this code works for me.

mdboom commented on setupext.py:
```diff
@@ -1817,8 +1811,6 @@ def add_flags(self, ext):
              tk_lib_dir, tk_inc_dir, tk_lib) = result
             ext.include_dirs.extend([tcl_inc_dir, tk_inc_dir])
-            ext.library_dirs.extend([tcl_lib_dir, tk_lib_dir])
-            ext.libraries.extend([tcl_lib, tk_lib])
```
Do we need any of this stuff (including finding header files) anymore if we've duplicated the function signatures (as function pointer signatures) in our own code now anyway?

matthew-brett replied:
Well, we still need the definitions for these data types:

```
    Tcl_Interp
    Tcl_Command
    Tk_PhotoHandle
    Tk_PhotoImageBlock
    Tk_Window
    Tcl_CmdProc
    ClientData
    Tcl_CmdDeleteProc
```

I guess we could copy those out of some TCL / Tk headers and then be entirely free of the system TCL / Tk.

matthew-brett replied:
I adding a minimal set of defines from TCL / Tk in the latest commit.  It seems to compile and run correctly on OSX, Linux and Windows, but I would appreciate some eyes on the minimal header, because of my aforementioned C-fu weakness.

If this new header is OK, then I'll go ahead and delete all the TCL / Tk finding logic in `setupext.py`.

matthew-brett commented:
I did some timing on the tcl / tk symbol search on Windows.  It's very fast.   Package / function to replicate the search at https://github.com/matthew-brett/tktest: 

```
In [1]: import numpy
In [2]: from scipy import io, stats, ndimage, optimize, sparse
In [3]: from sklearn import *
In [4]: from skimage import *
In [5]: from tktest import search_for_tk
In [6]: timeit search_for_tk()
10000 loops, best of 3: 106 µs per loop
```

matthew-brett commented:
Test failures seem to be due to freetype source server, not these changes.

matthew-brett commented:
OK - I checked current Tcl / Tk trunk, and I've edited the header excerpts, so they are compatible from Tcl / Tk 8.4 through current Tcl / Tk trunk.  I think we should get away with these header excerpts for a long time.

I've removed all the Tcl / Tk discovery stuff as no longer needed, and tested on OSX, Linux and Windows.  I think this one is ready to merge.

matthew-brett commented:
Github seems to be having trouble today - the latest commit for this branch is de99a76 - I don't see that yet in this interface.   When it arrives, you should have the changes I referred to in my last message.

tacaswell commented:
:+1: on this getting backported to 1.5.2rc1.  As soon as that is done I will do an rc2, still with the target of a final next weekend.

efiring commented:
Travis passed.  Appveyor stalled in the build stage on the Py3.5 job.  I didn't see anything indicating the build failure was related to the PR, so I am closing and reopening to try again.

efiring commented:
Well, I don't have any idea what the problem is, but Appveyor failed to build on Python 3.5 just like last time.  Very little output is generated.  It seems to stall before actually doing any conda installations.

matthew-brett commented:
Appveyor now passing on all Pythons for some reason.   Travis failure looks unrelated.    Appveyor and Travis tests now testing tkagg import and succeeding.   Tests of this branch on OSX at [MacPython/matplotlib-wheels](https://travis-ci.org/MacPython/matplotlib-wheels/builds/132604034) finding and importing tkagg correctly (unrelated failures but tests passing on Pythons 2.7 3.4 3.5).

efiring commented:
Everything passed and Appveyor showed the tkagg test result correctly, but it looks like Travis either didn't execute the tkagg import test, or swallowed its output:
```
The command "echo Testing import of tkagg backend
MPLBACKEND="tkagg" python -c 'import matplotlib.pyplot as plt; print(plt.get_backend())'
echo Testing using $NPROC processes
```

matthew-brett commented:
OSX builds [now green](https://travis-ci.org/MacPython/matplotlib-wheels/builds/132615025) - successful tkagg import checks for [Python 2.7](https://travis-ci.org/MacPython/matplotlib-wheels/builds/132615025) [Python 3.4](https://travis-ci.org/MacPython/matplotlib-wheels/jobs/132615033#L8674) [Python 3.5](https://travis-ci.org/MacPython/matplotlib-wheels/jobs/132615034#L8690).

matthew-brett commented:
Eric - I think the missing output is just because the command is in a block with other commands.  The output is further down after the whole command input block - e.g. https://travis-ci.org/matplotlib/matplotlib/jobs/132588520#L1018

efiring commented:
backported to v1.5.x as b80e0f1

cournape commented:
When updating matplotlib for canopy, I am seeing the error "Could not find TCL routines" with 1.5.3.

At first, I suspect an issue in how we build python for canopy, but I have been able to reproduce the issue with python 2.7.12 from python.org and just installing wheels from Gholke website. The failure appears for both 1.5.3 and 2.0.0.

I could also reproduce the issue w/ 1.5.3 using conda (2.7.13 this time)

cournape commented:
Hm, that should really be an issue on its own, sorry
